engine/consensus/approvals/approval_collector.go

```diff
@@ -0,0 +1,169 @@
+type ApprovalCollector struct {
+	incorporatedBlock                    *flow.Header                        // block that incorporates execution result
+	incorporatedResult                   *flow.IncorporatedResult            // incorporated result that is being sealed
+	chunkCollectors                      []*ChunkApprovalCollector           // slice of chunk collectorTree that is created on construction and doesn't change
+	aggregatedSignatures                 map[uint64]flow.AggregatedSignature // aggregated signature for each chunk
+	lock                                 sync.RWMutex                        // lock for modifying aggregatedSignatures
+	seals                                mempool.IncorporatedResultSeals     // holds candidate seals for incorporated results that have acquired sufficient approvals; candidate seals are constructed  without consideration of the sealability of parent results
+	numberOfChunks                       int                                 // number of chunks for execution result, remains constant
+	requiredApprovalsForSealConstruction uint                                // min number of approvals required for constructing a candidate seal
+}
+
+func NewApprovalCollector(result *flow.IncorporatedResult, incorporatedBlock *flow.Header, assignment *chunks.Assignment, seals mempool.IncorporatedResultSeals, requiredApprovalsForSealConstruction uint) *ApprovalCollector {
+	chunkCollectors := make([]*ChunkApprovalCollector, 0, result.Result.Chunks.Len())
+	for _, chunk := range result.Result.Chunks {
+		chunkAssignment := assignment.Verifiers(chunk).Lookup()
+		collector := NewChunkApprovalCollector(chunkAssignment, requiredApprovalsForSealConstruction)
+		chunkCollectors = append(chunkCollectors, collector)
+	}
+	return &ApprovalCollector{
+		incorporatedResult:                   result,
+		incorporatedBlock:                    incorporatedBlock,
+		numberOfChunks:                       result.Result.Chunks.Len(),
+		chunkCollectors:                      chunkCollectors,
+		requiredApprovalsForSealConstruction: requiredApprovalsForSealConstruction,
+		aggregatedSignatures:                 make(map[uint64]flow.AggregatedSignature, result.Result.Chunks.Len()),
+		seals:                                seals,
+	}
+}
+
+// IncorporatedBlockID returns the ID of block which incorporates execution result
+func (c *ApprovalCollector) IncorporatedBlockID() flow.Identifier {
+	return c.incorporatedResult.IncorporatedBlockID
+}
+
+// IncorporatedBlock returns the block which incorporates execution result
+func (c *ApprovalCollector) IncorporatedBlock() *flow.Header {
+	return c.incorporatedBlock
+}
+
+func (c *ApprovalCollector) chunkHasEnoughApprovals(chunkIndex uint64) bool {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	_, found := c.aggregatedSignatures[chunkIndex]
+	return found
+}
+
+func (c *ApprovalCollector) SealResult() error {
```

Comment:
It's imperative to ask ApprovalCollector to add the seal to the mempool, as if the caller assumes it must be able to generate a seal. 

We should turn this into a private method. 

And add another method `EmergencySeal() (*Seal, error)` that returns a seal if emergency sealing is triggered.

So that if emergency sealing is not triggered, seal will not be force generated. 

And since the condition of triggering emergency sealing is out of this module's control, it provides the `EmergencySeal` function for caller to decide it is needed.

EmergencySeal could also be cached, if it's possibly called multiple times.


I think we could let ApprovalCollector just return the seal, and let the caller put it to the seals mempool.

Reply:
I don't like this approach because currently `ApprovalCollector` puts everything into mempool, in your suggestion it will be split between two instances. Moreover, emergency sealing is a temporary measure and will be removed in future phases.